libevmasm/Assembly.cpp

```diff
@@ -784,6 +785,15 @@ std::map<u256, u256> const& Assembly::optimiseInternal(
 				return _i == AssemblyItem{Instruction::MSIZE} || _i.type() == VerbatimBytecode;
 			});
 
+			auto runGas = [&](AssemblyItems const& items, KnownState _state) {
+				GasMeter gasMeter{std::make_shared<KnownState>(_state), _settings.evmVersion};
+				GasMeter::GasConsumption gas;
+				for (auto const& item: items)
+					gas += gasMeter.estimateMax(item);
+				return gas;
+			};
```

Comment:
Your implementation ignores `GasConsumption.isInfinite` so it will give us wrong results if any of the items has an infinite estimate. If you're assuming it won't happen, you should at least have an assert against it.

There's a correct implementation in `Inliner.cpp`: https://github.com/ethereum/solidity/blob/ebdce269005e8f2a6652bad64a000725ff1624fe/libevmasm/Inliner.cpp#L47-L59

I think it would be to reuse it here too. It looks general enough that perhaps we could just move it to `GasMeter.h`.